### PR TITLE
[ossrdbms] Add server data encryption key management  commands for PostgreSQL and MySQL

### PR DESCRIPTION
**Pull request by shinilm**
**Description of PR (Mandatory)**  
OSSRDBMS enabled customers to bring own keys for data encryption.  This PR is to introduce tOssrdbms data encryption CLI commands for PostgreSQL and MySQL

**Testing Guide**  

PostgreSQL Server update/create with identity

az postgres server update --name <server name> -g <resoure_group> --assign_identity

az postgres server create --name <server name> -g <resoure_group> --location eastus --storage-size 51200 -u <userver Name> -p <passoword> --backup-retention 7 --sku-name gp_gen5_2 --geo-redundant-backup Disabled --assign_identity      

Set Wrap/UNWRAP/Get permissions on key vault for the server identiry
az keyvault set-policy --name <keyvault name> -g <resoure_group> --key-permissions get unwrapKey wrapKey --object-id <principl id of the server - 4e647c70-733b-4dbf-b7b2-0bcc537dc450> 

Set the key to the server
az postgres server key --name <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

Get the key
az postgres server key show --name <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

List the Key
az postgres server key list --name orcas-byok-tst -g shinim

Drop the Key
az postgres server key delete <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

===================================================================================================================================
MYSQL Server update/create with identity

az mysql server update --name <server name> -g <resoure_group> --assign_identity

az mysql server create --name <server name> -g <resoure_group> --location eastus --storage-size 51200 -u <userver Name> -p <passoword> --backup-retention 7 --sku-name gp_gen5_2 --geo-redundant-backup Disabled --assign_identity      

Set Wrap/UNWRAP/Get permissions on key vault for the server identiry
az keyvault set-policy --name <keyvault name> -g <resoure_group> --key-permissions get unwrapKey wrapKey --object-id <principl id of the server - 4e647c70-733b-4dbf-b7b2-0bcc537dc450> 

Set the key to the server
az mysql server key --name <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

Get the key
az mysql server key show --name <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

List the Key
az mysql server key list --name orcas-byok-tst -g shinim

Drop the Key
az mysql server key delete <server name> -g <resoure_group>  --kid <key url like https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901>

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 1]
```diff
@@ -32,13 +32,15 @@ def _complex_params(command_group):
             c.argument('geo_redundant_backup', arg_type=get_enum_type(['Enabled', 'Disabled']), options_list=['--geo-redundant-backup'], help='Enable or disable geo-redundant backups. Default value is Disabled. Not supported in Basic pricing tier.')
             c.argument('storage_mb', options_list=['--storage-size'], type=int, help='The storage capacity of the server (unit is megabytes). Minimum 5120 and increases in 1024 increments. Default is 51200.')
             c.argument('auto_grow', arg_type=get_enum_type(['Enabled', 'Disabled']), options_list=['--auto-grow'], help='Enable or disable autogrow of the storage. Default value is Enabled.')
+            c.argument('assign_identity', options_list=['--assign_identity'], help='Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault.')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py]: options_list element should be `--assign-identity`

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 2]
```diff
@@ -32,13 +32,15 @@ def _complex_params(command_group):
...
 
             c.argument('location', arg_type=get_location_type(self.cli_ctx), validator=get_default_location_from_resource_group)
             c.argument('version', help='Server major version.')
 
         with self.argument_context('{} server update'.format(command_group)) as c:
             c.ignore('family', 'capacity', 'tier')
             c.argument('sku_name', options_list=['--sku-name'], help='The name of the sku. Follows the convention {pricing tier}_{compute generation}_{vCores} in shorthand. Examples: B_Gen5_1, GP_Gen5_4, MO_Gen5_16.')
+            c.argument('assign_identity', options_list=['--assign_identity'], help='Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault.')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py]: The same as above. I will recommend you to abstract the duplicated argument description using `CLIArgumentType`.

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 3]
```diff
@@ -86,8 +88,12 @@ def _complex_params(command_group):
             c.argument('storage_mb', options_list=['--storage-size'], type=int, help='The storage capacity of the server (unit is megabytes). Minimum 5120 and increases in 1024 increments. Default is 51200.')
             c.argument('backup_retention', options_list=['--backup-retention'], type=int, help='The number of days a backup is retained. Range of 7 to 35 days. Default is 7 days.', validator=retention_validator)
             c.argument('auto_grow', arg_type=get_enum_type(['Enabled', 'Disabled']), options_list=['--auto-grow'], help='Enable or disable autogrow of the storage. Default value is Enabled.')
+            c.argument('assign_identity', options_list=['--assign_identity'], help='Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault.')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py]: The same.

File: src/azure-cli/azure/cli/command_modules/rdbms/_help.py [hunk 4]
```diff
@@ -689,6 +689,26 @@
 short-summary: List the private link resources supported for a MySQL server.
 """
 
+helps['mysql server encryption key'] = """
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_help.py]: Where is the command registration for `encyption key`? I only found `server key` related commands. Is it a typo?

File: src/azure-cli/azure/cli/command_modules/rdbms/custom.py [hunk 5]
```diff
@@ -459,4 +474,63 @@ def reject_private_endpoint_connection(cmd, client, resource_group_name, server_
     return _update_private_endpoint_connection_status(
         cmd, client, resource_group_name, server_name, private_endpoint_connection_name, is_approved=False,
         description=description)
+
+
+def server_key_create(client, resource_group_name, server_name, kid):
+
+    """Create Server Key."""
+
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/custom.py]: Should the key be in the same resource group with server?

**Comment by yonzhan** [comment]
add to S167

File: src/azure-cli/azure/cli/command_modules/rdbms/custom.py (hunk 5 above)
Comment (shinilm) [reply src/azure-cli/azure/cli/command_modules/rdbms/custom.py]: The actual Key creation happen in Azure keyvault and that is outside the scope of this commands.  that can be in any subscription/ resource group. Once the key is created in keyvualt the  key will be passed to postgres/mysql server to enable the byok.  The create operation here refers to updating the server with the key created in the key vault and enable encryption.  This is inline with th existing behaviour with the sql server

File: src/azure-cli/azure/cli/command_modules/rdbms/_help.py (hunk 4 above)
Comment (shinilm) [reply src/azure-cli/azure/cli/command_modules/rdbms/_help.py]: Fixed

**Comment by bim-msft** [comment]
@shinilm You may try out the following code:

```python
    @ResourceGroupPreparer()
    @ServerPreparer(engine_type='mysql')
    def test_mysql_proxy_resources_mgmt(self, resource_group, server, database_engine):
        self._test_firewall_mgmt(resource_group, server, database_engine)
        self._test_vnet_firewall_mgmt(resource_group, server, database_engine)
        self._test_db_mgmt(resource_group, server, database_engine)
        self._test_configuration_mgmt(resource_group, server, database_engine)
        self._test_log_file_mgmt(resource_group, server, database_engine)
        self._test_private_link_resource(resource_group, server, database_engine, 'mysqlServer')
        self._test_private_endpoint_connection(resource_group, server, database_engine)

        vault_name = self.create_random_name('testmysqlproxy', 24)
        self._test_data_encryption(resource_group, server, database_engine, vault_name)

    @ResourceGroupPreparer()
    @ServerPreparer(engine_type='postgres')
    def test_postgres_proxy_resources_mgmt(self, resource_group, server, database_engine):
        self._test_firewall_mgmt(resource_group, server, database_engine)
        self._test_vnet_firewall_mgmt(resource_group, server, database_engine)
        self._test_db_mgmt(resource_group, server, database_engine)
        self._test_configuration_mgmt(resource_group, server, database_engine)
        self._test_log_file_mgmt(resource_group, server, database_engine)
        self._test_private_link_resource(resource_group, server, database_engine, 'postgresqlServer')
        self._test_private_endpoint_connection(resource_group, server, database_engine)

        vault_name = self.create_random_name('testpgproxy', 24)
        self._test_data_encryption(resource_group, server, database_engine, vault_name)
```

```python
def _test_data_encryption(self, resource_group, server, database_engine, vault_name):
        resource_prefix = 'ossrdbmsbyok'
        key_name = self.create_random_name(resource_prefix, 32)

        # add identity to server
        server_resp = self.cmd('{} server update -g {} --name {} --assign-identity'
                               .format(database_engine, resource_group, server)).get_output_in_json()
        server_identity = server_resp['identity']['principalId']

        # create vault and acl server identity
        self.cmd('keyvault create -g {} -n {} --enable-soft-delete true --enable-purge-protection true'
                 .format(resource_group, vault_name))

        # create key
        key_resp = self.cmd('keyvault key create --name {} -p software --vault-name {}'
                            .format(key_name, vault_name)).get_output_in_json()

        self.cmd('keyvault set-policy -g {} -n {} --object-id {} --key-permissions wrapKey unwrapKey get list'
                 .format(resource_group, vault_name, server_identity))
```

File: src/azure-cli/azure/cli/command_modules/rdbms/tests/latest/test_rdbms_commands.py [hunk 6]
```diff
@@ -284,7 +284,7 @@ def test_postgres_proxy_resources_mgmt(self, resource_group, server, database_en
         self._test_log_file_mgmt(resource_group, server, database_engine)
         self._test_private_link_resource(resource_group, server, database_engine, 'postgresqlServer')
         self._test_private_endpoint_connection(resource_group, server, database_engine)
-        self._test_data_encryption(resource_group, server, database_engine)
+        self._test_data_encryption(resource_group, server, database_engine, self.create_random_name('ossrdbmsbyokpostgres', 24))
```
Comment (bim-msft) [inline src/azure-cli/azure/cli/command_modules/rdbms/tests/latest/test_rdbms_commands.py]: You may shorten the vault name and try it again.

File: src/azure-cli/azure/cli/command_modules/rdbms/_help.py [hunk 7]
```diff
@@ -689,6 +689,26 @@
...
+"""
+
+helps['mysql server key delete'] = """
+type: command
+short-summary: Deletes a server key.
+"""
+
+helps['mysql server key show'] = """
+type: command
+short-summary: Shows a server key.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_help.py]: use active voice for description. e.g. Creates ->Create

File: src/azure-cli/azure/cli/command_modules/rdbms/_help.py [hunk 8]
```diff
@@ -1026,3 +1046,23 @@
...
+"""
+
+helps['postgres server key delete'] = """
+type: command
+short-summary: Deletes a server key.
+"""
+
+helps['postgres server key show'] = """
+type: command
+short-summary: Shows a server key.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_help.py]: The same as above

File: src/azure-cli/azure/cli/command_modules/rdbms/commands.py [hunk 9]
```diff
@@ -388,3 +402,19 @@ def load_command_table(self, _):
                             postgres_private_link_resources_sdk,
                             client_factory=cf_postgres_private_link_resources_operations) as g:
         g.show_command('list', 'list_by_server')
+
+    with self.command_group('mysql server key',
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/commands.py]: can we mark the new command group as Preview for one or two sprint to test the behaviors for the commands?

File: src/azure-cli/azure/cli/command_modules/rdbms/commands.py [hunk 10]
```diff
@@ -388,3 +402,19 @@ def load_command_table(self, _):
...
+                            mysql_key_sdk,
+                            client_factory=cf_mysql_server_keys_operations) as g:
+        g.custom_command('create', 'server_key_create')
+        g.custom_command('delete', 'server_key_delete', confirmation=True)
+        g.custom_show_command('show', 'server_key_get')
+        g.command('list', 'list')
+
+    with self.command_group('postgres server key',
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/commands.py]: The same for this command group.

**Pull request merged**